FEATURE: Add `discourse_theme rspec <path to spec files>` command

tgxworld opened the pull request:
What does this change do?

This commit adds the `rspec` command to the `discourse_theme` CLI so
that themes can write system specs and run them easily without the need
to set up a local Discourse development environment.

Example:

```
discourse_theme rspec /path/to/theme
```

or

```
discourse_theme rspec /path/to/theme/spec/specific-folder
```

or

```
discourse_theme rspec /path/to/theme/spec/specific_rspec_file.rb
```

or

```
discourse_theme rspec
/path/to/theme/spec/specific_rspec_file_with_line_number.rb:3
```

Under the hood, the command runs a Docker container using the
`discourse/discourse_test:release` image and sets up the test
environment. This does mean the first run will incur some overhead and
there are ways to reduce this overhead but I will address them in
subsequent PRs. Once the Docker container is running, it simply runs
`bundle exec rspec` in the Docker container for the path that is given.

When running system specs, the `--headless` option can also be used to
run the system specs in headless mode. While we can't run a non-headless
chrome browser in the container, we can run a chrome browser locally and ask the
Rails system test to connect to the local browser. This is done by
having the CLI start a `chromedriver` instance locally and then passing
the `CAPYBARA_SERVER_HOST` and `CAPYBARA_REMOTE_DRIVER_URL` env
variables when running the RSpec command.

tgxworld committed:
FEATURE: Add `discourse_theme rspec <path to spec files>` command

What does this change do?

This commit adds the `rspec` command to the `discourse_theme` CLI so
that themes can write system specs and run them easily without the need
to set up a local Discourse development environment.

Example:

```
discourse_theme rspec /path/to/theme
```

or

```
discourse_theme rspec /path/to/theme/spec/specific-folder
```

or

```
discourse_theme rspec /path/to/theme/spec/specific_rspec_file.rb
```

or

```
discourse_theme rspec
/path/to/theme/spec/specific_rspec_file_with_line_number.rb:3
```

Under the hood, the command runs a Docker container using the
`discourse/discourse_test:release` image and sets up the test
environment. This does mean the first run will incur some overhead and
there are ways to reduce this overhead but I will address them in
subsequent PRs. Once the Docker container is running, it simply runs
`bundle exec rspec` in the Docker container for the path that is given.

When running system specs, the `--headless` option can also be used to
run the system specs in headless mode. While we can't run a non-headless
chrome in the browser, we can run a chrome browser locally and ask the
Rails system test to connect to the localy browser. This is done by
having the CLI start a `chromedriver` instance locally and then passing
the `CAPYBARA_SERVER_HOST` and `CAPYBARA_REMOTE_DRIVER_URL` env
variables when running the RSpec command.

tgxworld commented on lib/discourse_theme/cli.rb:
```diff
@@ -167,6 +183,94 @@ def run(args)
+          WebDriver.start(browser: :chrome)
+
+          rspec_envs.push("SELENIUM_HEADLESS=0")
+          rspec_envs.push("CAPYBARA_SERVER_HOST=0.0.0.0")
+          rspec_envs.push("CAPYBARA_REMOTE_DRIVER_URL=http://host.docker.internal:9515")
```
See https://github.com/discourse/discourse/blob/6e2b484f12338a92642dc6ccc982c5b4b6d5cf58/spec/rails_helper.rb#L387-L390

nbianca commented on lib/discourse_theme/cli.rb:
```diff
@@ -167,6 +183,94 @@ def run(args)
+        raise DiscourseTheme::ThemeError.new "'#{dir} does not exist" unless Dir.exist?(dir)
+
+        # Checks if the container is running
+        container_name = DISCOURSE_TEST_DOCKER_CONTAINER_NAME
```
I am wondering if this should contain the name of the theme to allow for multiple theme specs to run simultaneously. However, this complicates stuff.

tgxworld replied:
This is resolved in https://github.com/discourse/discourse_theme/pull/28/commits/2a0c8d9fc7f7e30cd6da14ef1365575ad20c9319 as the Docker container is just a generic test env container and not tied directly to any theme.

davidtaylorhq commented on lib/discourse_theme/cli.rb:
```diff
@@ -167,6 +183,94 @@ def run(args)
+                --entrypoint=/sbin/boot \
+                --name=#{container_name} \
+                -v #{dir}:/tmp/#{basename} \
+                discourse/discourse_test:release
```
`docker run {blah}` does not guarantee to use the latest version of the image. If the workstation already has an old copy, that will be used instead.

So I think the best strategy is probably to do what our `./launcher` script does in production: run `docker pull discourse/discourse_test:release` before the `docker run`

tgxworld replied:
This is a good point, I do feel like running `docker pull` is going to add significant overhead so I think I'll go with a versioned approach. In other words, for this version of the gem we will run this version of the `discourse/discourse_test:release` image.

tgxworld replied:
Btw I intend to follow up with this in another PR while I figure out if we have versioning for our images.

davidtaylorhq replied:
We don't version our discourse_test image (tags are [here](https://hub.docker.com/r/discourse/discourse_test/tags))

The good news is that `docker pull` is basically free if you already have the latest image. It compares the image hashes of the local/remote versions before doing anything.

```
❯ time docker pull discourse/discourse_test:release
release: Pulling from discourse/discourse_test
Digest: sha256:2a78135423931fe11fd7edfccf19c476125a1fbc0e12cabf56cf79c2405877f8
Status: Image is up to date for discourse/discourse_test:release
docker.io/discourse/discourse_test:release
docker pull discourse/discourse_test:release

0.08s user
0.06s system
10% cpu
1.409 total
```

I don't think we should 'release' this without the `docker pull`. If we do, we're gonna have a world of pain when people start reporting weird issues caused by out-of-date images.

tgxworld replied:
I've added `--pull=always` when starting the container to resolve this :+1:

nbianca commented on lib/discourse_theme/cli.rb:
```diff
@@ -167,6 +183,94 @@ def run(args)
+
+          execute(
+            command: <<~CMD.squeeze(" "),
+              docker run -d \
```
I wonder if we should add `--rm` to this so that we do not have to worry about doing `docker rm` too above.

tgxworld replied:
Good point :+1: I changed this in https://github.com/discourse/discourse_theme/pull/28/commits/4b3ae2c581236974b0e19630f865b16ae3abeb8b

tgxworld replied:
Actually this made the scripting harder because we end up having to poll to figure out if the Docker container has been removed by Docker. Explicitly running `docker rm` is much easier.

davidtaylorhq commented on lib/discourse_theme/cli.rb:
```diff
@@ -167,6 +183,94 @@ def run(args)
+        verbose = !!args.delete("--verbose")
+        headless = !!args.delete("--headless")
+
+        if !is_running || args.delete("--rebuild")
```
What if a container is already running, but it has a volume mount to a different theme? Can we detect that and stop/start it?

tgxworld replied:
This should be resolved in https://github.com/discourse/discourse_theme/pull/28/commits/2a0c8d9fc7f7e30cd6da14ef1365575ad20c9319 since the volume mount is no longer tied directly to a theme.

nbianca commented on lib/discourse_theme/cli.rb:
```diff
@@ -167,6 +183,94 @@ def run(args)
+          execute(
+            command: <<~CMD.squeeze(" "),
+              docker run -d \
+                -p 31337:31337 \
```
This port should be randomized if we want to run multiple theme specs at the same time (see previous comment).

tgxworld replied:
Hmm this is tricky because the ports have to be exposed before hand and we are not creating a new container each time we need to run the system tests for a theme. For now, I think we should just not support running multiple theme specs at the same time to keep this simple.

nbianca commented on lib/discourse_theme/cli.rb:
```diff
@@ -167,6 +183,94 @@ def run(args)
+          execute(
+            command:
+              "docker exec -u discourse:discourse #{container_name} ruby script/docker_test.rb --no-tests --checkout-ref origin/tests-passed",
+            message: "Checking out latest Discourse source code...",
```
I think it would be fantastic if we would be able to test with other Discourse versions too in order to make sure that `.discourse_compatibility` is correct. However, we can leave this for the future.

nbianca commented on lib/discourse_theme/cli.rb:
```diff
@@ -167,6 +183,94 @@ def run(args)
+
+        execute(
+          command:
+            "docker exec #{rspec_envs} -u discourse:discourse #{container_name} bundle exec rspec --color --tty #{File.join("/tmp", basename, spec_path)}".squeeze(
```
I wonder if a `-t` is missing from `docker exec` and that is why you must pass `--color` and `--tty` to rspec.

tgxworld replied:
O thanks! This was it :) I have made the change.

davidtaylorhq commented on lib/discourse_theme/cli.rb:
```diff
@@ -167,6 +183,94 @@ def run(args)
+
+        if !(
+             output =
+               execute(command: "docker ps -a --format '{{json .}}' | grep #{container_name}")
```
If the container doesn't exist, the `grep` returns a non-zero exit status and I get

```
✘ Error occured while running: `docker ps -a --format '{{json .}}' | grep discourse_theme_test`:
```
I can work past it by adding `exit_on_error: false`, but then that'll swallow up any legit docker errors. 

Maybe we just run the docker command, and then parse the JSON in ruby?

tgxworld replied:
This might be OSX specific because on linux I get an exit status of `0` even if the container doesn't exist. 

> Maybe we just run the docker command, and then parse the JSON in ruby?

Yup I think we'll have to go this way to support multiple platform.

tgxworld replied:
O `docker ps` has a `--filter` option so I went with that instead in https://github.com/discourse/discourse_theme/pull/28/commits/0d79a16d3dd167a30a267cd2d8340e41850e9e84.

nbianca commented on test/test_cli.rb:
```diff
@@ -269,4 +269,212 @@ def test_new
+      end
+
+      cli.expects(:execute).with(
+        command: <<~COMMAND.squeeze(" "),
```
Seeing how often `squeeze` is used for the command argument, I wonder if it makes sense to move it to `execute`. I believe this is used only when printing the command line anyway, right? I mean, when executing the command it does not matter if there are some extra spaces.

tgxworld replied:
Good point :+1: I'm using squeeze mainly to make the mocks in the tests easier to write. I've moved `squeeze` into the `execute` method.

davidtaylorhq commented on lib/discourse_theme/cli.rb:
```diff
@@ -12,14 +19,23 @@ def self.settings_file=(filename)
+        download DIR          - Downloads a theme from the server and stores it in the specified directory.
+        upload DIR            - Uploads the theme from the specified directory to Discourse.
+        watch DIR             - Watches the theme in the specified directory and synchronizes any changes with Discourse.
+        rspec DIR [OPTIONS]   - Runs the RSpec tests in the specified directory.
```
How easy would it be to provide an option to run against an existing local development environment? I think most CDCK devs would prefer that, and it would also unlock running theme system specs against discourse installations with certain plugins. 🤔

davidtaylorhq replied:
Another thing I'm realising while trying this out: the discourse_test image is sloooooow when running on Apple Silicon with architecture virtualisation. Being able to run against a regular dev environment would be especially useful here!

(or we try and build an ARM version of the discourse_test image... but that's a whole project in itself)

tgxworld replied:
> How easy would it be to provide an option to run against an existing local development environment?

In theory super easy because all you really need to do is to just `bundle exec rspec` in the `discourse` repository and then just point it at the path of the theme's spec folder. I'll see if I integrate this into the `discourse_theme rspec` command. We might have to set some sort of ENV variable in order to tell the `discourse_theme` CLI where the discourse repository is located though.

davidtaylorhq commented on lib/discourse_theme/cli.rb:
```diff
@@ -167,6 +183,94 @@ def run(args)
+          execute(
+            command:
+              "docker exec -u discourse:discourse #{container_name} bundle exec rake docker:test:setup",
+            message: "Setting up Discourse test environment...",
```
This takes a very long time on my machine (probably because it's virtualising an x86 container on an M1 mac). Would be nice if there was some kind of feedback on what it's doing? (I guess stuff like bundle install, db create/migrate, etc?)

davidtaylorhq replied:
Ah I see there is a verbose flag... so maybe that's good enough for now. Still would be a 'nice to have' if it could print the main milestones in non-verbose mode.

tgxworld committed:
Use `--filter name=discourse_theme_test` instead of grep for `docker ps`

tgxworld committed:
Allocate a tty to remove `--color` and `--tty` from rspec command

tgxworld committed:
Allow running system tests for multiple themes without having to rebuild

Before this commit, I was mounting the theme directory directing as a
volume mount to the docker container. However, volume mounts can only be
created when running a container so if the user wants to run tests for a
theme, we will have to stop the old container and start a new one and
setup the container all over again.

Instead of mounting the theme's directory directly into the container,
we instead just mount a tmp directory and copy the theme's spec folder
into the volumn mount when running the specs. This way, the volume mount
is not tied directly into a theme and the docker container can be used
to run the RSpec tests for any theme.

tgxworld committed:
Avoid setting up multisite to reduce setup time

tgxworld commented:
@davidtaylorhq @nbianca Can you have another look? You can see the new changes I made by following the commits I made to this PR :) Thank you!

tgxworld committed:
Remove stray comment

tgxworld committed:
Set `--allowed-ips` and `--allowed-origins` for chromedriver

Previously we were letting everything through which may pose a security
issue if user's local network is exposed.

nbianca reviewed (approved):
Fantastic work! I am looking forward to using this. 🤩

davidtaylorhq commented on lib/discourse_theme/cli.rb:
```diff
@@ -167,6 +184,118 @@ def run(args)
+
+          execute(
+            command: <<~CMD.squeeze(" "),
+              docker run -d \
```
This will spin up a background container, which can be quite expensive in terms of resources. I think we need to give the user an easy way to shut it down when they're done. 

In our standard docker dev environment, this is handled by `d/shutdown-dev`.

tgxworld replied:
Noted here I will add a `discourse_theme rspec --shutdown` option in a follow up PR.

tgxworld committed:
Pull latest image before starting container

tgxworld committed:
Add gem's version to container name

This is to allow us to identify which gem version that docker container
was created with allowing for easier debugging and also allows us to
recreate the container when the gem version is bumped.

tgxworld commented:
@davidtaylorhq @nbianca Thank you for the detailed review here. I've merged the changes in for now before the PR gets too big but will continue to ship smaller PRs to improve the CLI.
